### PR TITLE
RATIS-1174. Test leader change when close stream

### PR DESCRIPTION
**Pull request by runzhiwang**
## What changes were proposed in this pull request?

Test leader change when close stream

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-1174

## How was this patch tested?

new ut

### CODE REVIEW

**Comment by runzhiwang** [comment]
This pr depends on https://github.com/apache/incubator-ratis/pull/308

File: ratis-test/src/test/java/org/apache/ratis/datastream/DataStreamAsyncClusterTests.java [hunk 1]
```diff
@@ -69,12 +69,29 @@ public void testMultipleStreamsMultipleServers() throws Exception {
...
+  }
+
+  void runTestDataStream(CLUSTER cluster, boolean stepDownLeader) throws Exception {
     RaftTestUtil.waitForLeader(cluster);
 
     final List<CompletableFuture<Long>> futures = new ArrayList<>();
-    futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 5, 10, 1_000_000, 10), executor));
-    futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 2, 20, 1_000, 10_000), executor));
+    if (stepDownLeader) {
+      futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 1, 1, 1_000_000, 10, stepDownLeader), executor));
```
Comment (szetszwo) [inline ratis-test/src/test/java/org/apache/ratis/datastream/DataStreamAsyncClusterTests.java:90]: Tried to run with more streams but the test may fail once in a while.
```
@@ -87,7 +87,7 @@ public abstract class DataStreamAsyncClusterTests<CLUSTER extends MiniRaftCluste
 
     final List<CompletableFuture<Long>> futures = new ArrayList<>();
     if (stepDownLeader) {
-      futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 1, 1, 1_000_000, 10, stepDownLeader), executor));
+      futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 2, 10, 1_000, 10_000, stepDownLeader), executor));
     } else {
       futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 5, 10, 1_000_000, 10, stepDownLeader), executor));
       futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 2, 20, 1_000, 10_000, stepDownLeader), executor));
```

**Review by szetszwo — COMMENTED** [review]
Thanks a lot for working on this.  For leader stepping down, it is hard to check the server id.  Let's skip it?
```
@@ -366,10 +366,13 @@ public interface DataStreamTestUtils {
   }
 
   static void assertRaftClientMessage(
-      RaftClientMessage expected, RaftPeerId expectedServerId, RaftClientMessage computed, ClientId expectedClientId) {
+      RaftClientMessage expected, RaftPeerId expectedServerId, RaftClientMessage computed, ClientId expectedClientId,
+      boolean stepDownLeader) {
     Assert.assertNotNull(computed);
     Assert.assertEquals(expectedClientId, computed.getClientId());
-    Assert.assertEquals(Optional.ofNullable(expectedServerId).orElseGet(expected::getServerId), computed.getServerId());
+    if (!stepDownLeader) {
+      Assert.assertEquals(Optional.ofNullable(expectedServerId).orElseGet(expected::getServerId), computed.getServerId());
+    }
     Assert.assertEquals(expected.getRaftGroupId(), computed.getRaftGroupId());
   }
```
 Also, the log entry may become a different object.  Let's use assertEquals(..).
```
@@ -401,6 +404,6 @@ public interface DataStreamTestUtils {
 
     final LogEntryProto entryFromLog = searchLogEntry(ClientInvocationId.valueOf(request),
         RaftServerTestUtil.getRaftLog(division));
-    Assert.assertSame(entryFromStream, entryFromLog);
+    Assert.assertEquals(entryFromStream, entryFromLog);
   }
 }
```

File: ratis-test/src/test/java/org/apache/ratis/datastream/DataStreamAsyncClusterTests.java (hunk 1 above)
Comment (runzhiwang) [reply ratis-test/src/test/java/org/apache/ratis/datastream/DataStreamAsyncClusterTests.java:90]: @szetszwo The failed reason seems related with Isolate and deIsolate too many times. The root cause is hard to find. I will continue to find it.

Comment (szetszwo) [reply ratis-test/src/test/java/org/apache/ratis/datastream/DataStreamAsyncClusterTests.java:90]: I see. Isolate and deIsolate are not designed to run in multiple threads.  We should change leader in a separated thread.

Comment (szetszwo) [reply ratis-test/src/test/java/org/apache/ratis/datastream/DataStreamAsyncClusterTests.java:90]: We may change leader after the first future has completed.
```
+++ b/ratis-test/src/test/java/org/apache/ratis/datastream/DataStreamAsyncClusterTests.java
@@ -86,17 +86,25 @@ public abstract class DataStreamAsyncClusterTests<CLUSTER extends MiniRaftCluste
     RaftTestUtil.waitForLeader(cluster);
 
     final List<CompletableFuture<Long>> futures = new ArrayList<>();
-    if (stepDownLeader) {
-      futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 1, 1, 1_000_000, 10, stepDownLeader), executor));
-    } else {
-      futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 5, 10, 1_000_000, 10, stepDownLeader), executor));
-      futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 2, 20, 1_000, 10_000, stepDownLeader), executor));
-    }
+    futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 5, 10, 1_000_000, 10, false), executor));
+    futures.add(CompletableFuture.supplyAsync(() -> runTestDataStream(cluster, 2, 20, 1_000, 10_000, false), executor));
     final long maxIndex = futures.stream()
         .map(CompletableFuture::join)
         .max(Long::compareTo)
         .orElseThrow(IllegalStateException::new);
 
+    if (stepDownLeader) {
+      final RaftPeerId oldLeader = cluster.getLeader().getId();
+      final CompletableFuture<RaftPeerId> changeLeader = futures.get(0).thenApplyAsync(dummy -> {
+        try {
+          return RaftTestUtil.changeLeader(cluster, oldLeader);
+        } catch (Exception e) {
+          throw new CompletionException("Failed to change leader from " + oldLeader, e);
+        }
+      });
+      LOG.info("Changed leader from {} to {}", oldLeader, changeLeader.join());
+    }
+
```

**Pull request merged**